### ISSUE TITLE
[D8][WP][UX] Better workflow states and content save options.

### ISSUE THREAD

**Issue by klonos** [backdrop/backdrop-issues#1475]
This might actually be split to quite a few separate issues (some might have already been filed), but before we go ahead and do that, let's draft a plan here...

Here's what we have in Backdrop currently:
![backdrop-save_node_options](https://cloud.githubusercontent.com/assets/2423362/12071078/5da85de0-b0ec-11e5-8a9f-de42795c415b.png)

So basically only save/publish/cancel options. Why do we keep the "Published" option in a vtab? Couldn't we simply move this checkbox in the form of a separate "Publish" button (and change the "Save" button to something like "Save as draft")?

In [D7UX: "Save draft" and "Publish" buttons on node forms [#282122]](https://www.drupal.org/node/282122), it was suggested that there should be 3 buttons in the node edit/save form and 3 different scenarios:

On a new node:
[Publish] [Save as Draft] [Preview]

On a non-published node:
[Publish] [Save changes] [Preview]

On a published node:
[Save and republish] [Save and unpublish] [Preview]

(we removed the "Preview" option in #218)

That issue was closed as a duplicate of [Support revisions in different states [#218755]](https://www.drupal.org/node/218755) which aimed to introduce a new API but no UI changes. That seems to be committed/fixed and then there was [Support arbitrary workflow states [#1777388]](https://www.drupal.org/node/1777388) as a follow-up issue.

Here's what's was finally implemented in D8 (makes this issue here part of #378):
![d8-save_node_options](https://cloud.githubusercontent.com/assets/2423362/12071116/71ef5bf8-b0ee-11e5-97e7-cb67b3611f9c.png)

...and this is the publishing options in WordPress for its content create and edit forms (makes this issue also part of #1388):
![wp-publish_settings-add_page](https://cloud.githubusercontent.com/assets/2423362/12071147/5307d9ac-b0f0-11e5-8698-928cc0b81993.png)
![wp-publish_settings-add_page-save_as_pending_review](https://cloud.githubusercontent.com/assets/2423362/12071148/57d74224-b0f0-11e5-9ee6-758291ab2ce8.png)
![wp-publish_settings-edit_page](https://cloud.githubusercontent.com/assets/2423362/12071149/5e578262-b0f0-11e5-8604-3deef59bb283.png)

---
~PR by @docwilmot: https://github.com/backdrop/backdrop/pull/2943~
~PR by @jenlampton https://github.com/backdrop/backdrop/pull/3032~
Temporary PR by @docwilmot: https://github.com/backdrop/backdrop/pull/3051

**Comment by sutibun** [comment]
> So basically only save/publish/cancel options. Why do we keep the "Published" option in a vtab?

Way ahead of you. *Lots* of common options are "buried" in vertical tabs. In my mockup, pulled them out of there

https://github.com/backdrop/backdrop-issues/issues/1299#issuecomment-150751550

**Comment by laryn** [comment]
Perhaps relevant: 
[Workbench Moderation](https://www.drupal.org/project/workbench_moderation) is now in Drupal 8 core:

> Since Drupal 8.2.x this module is in core as "Content Moderation". However it is an experimental module and therefore in an alpha stage, prone to API breaking changing. We hope to provide an upgrade path from Workbench Moderation to Content Moderation. Content Moderation stores moderation states differently, so can be uninstalled, please take a look and submit bugs / feedback via the core issue queue.

**Comment by quicksketch** [comment]
Things have changed a bit now that Scheduling has been added:

![image](https://user-images.githubusercontent.com/100206/40511580-058258b8-5f55-11e8-85a3-47b18132017a.png)

It's also worth noting that Drupal 8 got rid of their dropbutton on the node form, as it was too confusing for users as to how to publish content.

**Comment by jenlampton** [comment]
I wonder how much of this would be solved by https://github.com/backdrop/backdrop-issues/issues/3157

**Comment by docwilmot** [comment]
I've started a PR that may help address this issue. It is a mashup of parts of the following Drupal issues:
[Create an interface for revisionable entities](https://www.drupal.org/project/drupal/issues/1612014)
[Support revisions in different states](https://www.drupal.org/project/drupal/issues/218755)
[Provide a better UX for creating, editing & managing draft revisions](https://www.drupal.org/project/drupal/issues/1776796)

A synopsis of what going on here:
D7/Backdrop allows revisions to be saved that are **older** copies of a node. But the node visible at `node/nid` will always be the latest revision . This is an issue if, as in the typical example, you have an "About us" page, and your company wants to edit it. It would be nice to allow editors to create **newer** revisions (drafts) of a node while not affecting the public facing node content at `node/nid`.

This PR allows you to set a revision as a Default revision, and will be the one seen when navigating to `node/nid`; a revision that is created that is not set as default will not be seen at `node/nid` but will show up in the list of revisions as a more recent draft. You can make any number of revisions before you decide to finally make one of them the actual Node at `node/nid`

This PR distills several D& issues, hundreds of comments and a lot of discussion. Kudos as usual to the D7 teams for the borrowed code. This PR doesnt bother with the discussion about what to call Drafts vs Revisions vs Etc Etc.

Every revision is made editable. On the node form, checking the "Make default" box will make the node the new default (not sure if this is yet foolproof, and could probably use a confirm form). Any revision is still not visible until the "published" checkbox is checked. We'd need to hide the scheduling ability to be only for the Default revision, or decide if scheduling should make a non-default revision default.

Finally we will need to reconsider the terminology of the existing "Save draft" checkbox since a forward revision is typically, elsewhere, known as a draft!

**Comment by docwilmot** [comment]
What motivated this PR was a comment by @klonos about allowing nodes/things to be sent to a "recycle bin" type state rather than deleting. This PR though doesnt go that far. As I see it, this PR adds "Revisioning module" or "Workbench" module to core. But to go further still, as far as allowing the recycle suggestion, would require, functionality such as "Workflow" or "Workbench Moderation" modules.

As the OP says WP has good workflow states support in their core product.

How far do we want to go with this PR though? What does the 80% rule suggest? DO we need this PR in core? Do we need the more advance moderation modules mentioned above? 

@jenlampton what you think?

**Comment by stpaultim** [comment]
I take that this PR is a proof of concept and that @docwilmot is just looking for feedback at this stage. Here are my initial reactions:

I took a little time tonight to look at the PR and feel like this has potential to be a really useful feature. This could be very exciting. Clearly, the UI is very confusing at this stage and I have not yet had time to make constructive suggestions. 

In terms of the 80% criteria. I think that this could be useful to 80% of users IF we make the UI really easy to use. That does not mean that 80% of users would actually make use of it, but I think that 80% of users might look at it and say that having it available in core is a big plus for Backdrop CMS. 

As it stands now, I don't think it would be that useful to most users (but only because it's not finished yet). 

I created a couple of screen shots to help people see what is going on so far in this PR:

![Edit_Page_Test___PR_2943_backdrop_backdrop_testing_site](https://user-images.githubusercontent.com/3144571/67629604-0b26c480-f846-11e9-9288-87fc07c8fcbd.png)

I would be included to call these something other than 'revisions' if we go this route. Off hand, calling them all 'drafts' makes more sense, but there might be better ideas. I just think that 'revisions' does not work so well with this way of managing revisions/drafts.

![Revisions_for_Test___PR_2943_backdrop_backdrop_testing_site](https://user-images.githubusercontent.com/3144571/67629656-d9fac400-f846-11e9-8767-1997d27790ef.png)

I think that a next step MIGHT be to make it clear on this page which is the current DEFAULT revision. That would help me a lot. 

![Revision_of_Test_from_Sun__10_27_2019_-_4_03am___PR_2943_backdrop_backdrop_testing_site](https://user-images.githubusercontent.com/3144571/67629626-7375a600-f846-11e9-9b79-433bf1e69315.png)

Any revision is editable. Any revision that is not the default revision, in this iteration is considered a draft. 

I would agree, that this is like the simplest version of a Workbench or Workflow type feature. I think this could be a great step in that direction. 

My only concern would be whether or not there would be an advantage of scoping out a more robust workflow module before committing to this. But, I'm not sure if that is necessary.

**Comment by docwilmot** [comment]
> I think that a next step MIGHT be to make it clear on this page which is the current DEFAULT revision. That would help me a lot.

Its the row highlighted yellow actually, but some text would be necessary, I agree, maybe after the author name, in brackets or something.

**Comment by docwilmot** [comment]
To allow full moderation, nodes would need assigned statuses. Curious thought though: node status in Backdrop is a Boolean field on the `node` table. The corresponding in WP seems to be [post_status](https://wordpress.org/support/article/post-status/) and WP comes with 8 predefined statuses. Would it be advisable you think to change the node status database field to text?
@quicksketch @herbdool @serundeputy what do you think?

**Comment by herbdool** [comment]
I'd be more inclined to make an additional column for this workflow status and keep the original backwards compatible. Then the new one could work in tandem with the other. A bit confusing but so many things rely on node status that I don't think we can break.

**Comment by docwilmot** [comment]
I'm thinking this may need to be done in steps. May best commit this as is, just as revisionable Entities first, then work on moderation after. 

We also have the same [problem with stub entities](https://www.drupal.org/project/drupal/issues/1551140) as faced in the D7 issue, which would need to be fixed first or as part of this patch.

**Comment by docwilmot** [comment]
Curiously the PR for fixing stub entities when merged with this PR, resulted in *more* test fails!

**Comment by klonos** [comment]
#4181 has been merged! 🎉 ...are you unblocked now @docwilmot?

...I have added the minor milestone candidate label to this issue, as #4181 has also been milestoned for that.

**Comment by docwilmot** [comment]
@klonos PR is green. Test away!

**Comment by docwilmot** [comment]
Thoughts/todo/to consider:
- A "set revision default" permission
- Currently there is no way to track from which revision you created a new revision. Would require a new DB field in the revisions table, like "parent_revision". That way the revisions could be threaded.
- There isn't any moderation yet. Would require either adding a new node status table or modifying the `status` column from boolean to integer.

**Comment by docwilmot** [comment]
I'm auto-advocating for this issue and may miss the meeting tomorrow, so update. This is passing tests so ready for review except for decisions about existing terminology.
- we call unpublished nodes "drafts" now; with this PR we may need to revisit that since we'll now be able to make true drafts. But how to call unpublished nodes then. Not sure really.
- should we allow scheduled publishing of revisions/drafts? Currently the scheduling function simply swithces a node from unpublished to published, which will work even if a revision is not the default revision, which would accomplish nothing much IMHO. Should scheduling only be for the default revision then? Or if we allow for revisions to have scheduled publishing: 
  - add a dependent field next to the scheduling radio if a revision is not default that says "promote to default" so that the revision is not just published but also made default at the scheduled time. 
  - Or should we just make it so that any scheduled non-default revision is just automatically made default (and add description text to make that obvious).

Finally once we finalize these UI options, will need to add a few more tests of the UI.

**Comment by jenlampton** [comment]
This is fantastic, I'm looking forward to testing it. 

I agree with @stpaultim that this is probably not an 80% use-case, but that it would be a good marketing tool. I'm not entirely sure how I feel about going in core right now, but after I test it I will probably get excited :) 

I'd love to hear what others think about the 80% criteria. Right now 0% of my Backdrop sites are using revisions or drafts -- but that's because it's not a nice experience to do so. But what if it were?!?

**Comment by quicksketch** [comment]
Could we update the summary of this issue at the top to reflect the approach being taken here? The original issue about the D8 "Save and publish" button seems very different from what we're now talking about, which seems to be primarily about forward, unpublished revisions.

> On the node form, checking the "Make default" box will make the node the new default (not sure if this is yet foolproof, and could probably use a confirm form). Any revision is still not visible until the "published" checkbox is checked. We'd need to hide the scheduling ability to be only for the Default revision, or decide if scheduling should make a non-default revision default. 

Although this may get tricky, my first impression is that this checkbox should not be needed. Using the current "Create new revision" checkbox along with the current "Draft/Published/Schedule" options seems like it should behave as follows:

- No new revision / Draft: If it is the active revision, unpublish the node entirely. If it is not the published revision, no effect. (This would be the default value when editing any revision except the active one).
- No new revision / Publish: Publish the revision being edited (make it the active revision).
- No new revision / Schedule: Make that revision active at the scheduled time. Remove scheduling for any other revisions.

- New revision / Draft: Make a new forward, non-active revision. Leave the current revision published.
- New revision / Publish: Make a new revision and make it be the published one for the node.
- New revision / Schedule: Leave the current revision (if any) published, publish the new revision at the schedule time. Remove scheduling for any other revisions.

All that said, I'm also not entirely sure this is core appropriate. I like the idea of making it easier to support, but perhaps this should start as a contrib module?

**Comment by docwilmot** [comment]
> All that said, I'm also not entirely sure this is core appropriate. I like the idea of making it easier to support, but perhaps this should start as a contrib module?

The difficulty with this as mentioned in [Support revisions in different states](https://www.drupal.org/project/drupal/issues/218755) is that core will always save the last revision as default, this is hard-coded in there. All the D7 moderation modules have to use workarounds to get that done. So that D7 issue was to make it so that there is a path forward for contrib to work with. 

What I've done here then, as per Backdrop's general tendency toward simplicity, is to do the same. The UI changes as you can see are minimal and are inevitable if we add this functionality (revisions will automatically be able to not be default, for example, so cant avoid that). I've left off actual moderation ability. 

I think if we want to have better contrib moderation, we'll need revisionable entities and the ability for forward revisions, in other words.

Could be wrong, so please read those veryy loong D7 issues mentioned in https://github.com/backdrop/backdrop-issues/issues/1475#issuecomment-544290251

**Comment by quicksketch** [comment]
If our object here is to just make things much easier for contrib, then perhaps we should add form element similar to your checkbox, but make it a `#type = 'value'` element? I think the concept of allowing a revision to be made without making it the default is simple, but the UI approaches could be varied and complicated.

**Comment by stpaultim** [comment]
I hesitate to comment without more time to articulate my thoughts, but under the theory that something is better than nothing, here are some quick thoughts.

1) Does the 80% use case mean that 80% actually use a feature or that 80% feel good knowing that the feature is there? I think this might meet the later.
2) I don't feel qualified to evaluate whether or not getting this into core will be helpful to other contrib modules. But, if that is the case, then that seems like a good argument. 
3) Even if the focus of this commit is on helping contrib and laying groundwork for future users, I think we should meet at least minimal usability issues for now. I have a few concerns:
* As usual, naming things is hard. I see the justification for "edit draft" tab and use of "default," but I found that bit confusing. In my testing, I quickly ended up with three new revisions that were not the "default," but it was not clear to me as editor if I was looking at the default or at a draft revision. * Also, if I click on "edit draft" does that take me to the most recent draft. I know from testing, that the "edit draft" tab is to edit the version you are looking at, but that was not clear to me. 
* Also, I tried saving one version that was NOT the default revision as a draft. When I viewed that revision, I got a big warning reminding me that that revision was a draft. But, when I view another revision that not default but was published, no indication of the status. 

I will try to think of actual recommendations later. 

#1) Should the "Edit Draft" actually say "Edit Draft #3" - be specific to which draft you are editing.
#2) Should there be a visual indication on any node page whenever you are looking at anything that is not the default version. 
#3) Should we refer to the "active" version of a node instead of the "default" version. Default just didn't work for me.

**Comment by docwilmot** [comment]
I dont know if this is what @quicksketch is suggesting (dont quite get the point re the `value` element), but we could make this an API improvement and hide it completely from the UI so everything works exactly as before, but then hope that contrib can come up with better implementation. 

This would be reasonable.

I personally feel that the existing changes in this PR are barely minimal compared to D8 or WP and could stay, but I'm not committed. Hoping for more opinions on how much if any of this we keep in the PR.

> Does the 80% use case mean that 80% actually use a feature or that 80% feel good knowing that the feature is there? I think this might meet the later.

Just to be clear, even if only 20% need moderation, this PR will help give it to them in contrib.

**Comment by docwilmot** [comment]
I'd like to somehow move forward on this, if possible, but I'm not sure if or why its stuck. Can we clearly say? 

It sounded like the UI change in the node form was an issue. If we remove the "Make default" checkbox from the node form, would that do?

Do we want to change revisions itself to a separate module?

Keeping in mind that D8 has Workflows and Content Moderation in core. This PR is nowhere near any of that.

**Comment by quicksketch** [comment]
> It sounded like the UI change in the node form was an issue. If we remove the "Make default" checkbox from the node form, would that do?

Yes, I think so. If you switch it from a checkbox to a `#type = value` then we keep the ability to make new, non-active revisions but it is not directly exposed as an option on the node form. Contrib could then set this value via a UI of it's choice.

> Do we want to change revisions itself to a separate module?

My initial reaction to this is that revisions should remain in the core UI. It's useful as it exists today, though not as useful as it could be. Revision functionality takes a huge amount of effort and affects almost every database table relating to content. If core didn't support it directly, I don't think contrib would be able to implement it at all.

**Comment by docwilmot** [comment]
I've reverted the node form UI change on my local, and was going to push but thought I'd ask for another clarification. This PR also changed the revisions list form: most important it adds the ability to edit any revision, and changes the wording from "Revert" which is a bit cryptic to me. Did we want to also remove all these changes from the PR? 
Old form:
![revisions old](https://user-images.githubusercontent.com/1042283/71546880-99174c80-2968-11ea-84b6-91dfd38af740.PNG)
New form (this PR):
![revisions new](https://user-images.githubusercontent.com/1042283/71546889-a3d1e180-2968-11ea-8a6d-77427c07d0d5.PNG)

**Comment by jenlampton** [comment]
> This PR also changed the revisions list form: most important it adds the ability to edit any revision, and changes the wording from "Revert" which is a bit cryptic to me. Did we want to also remove all these changes from the PR?

I think these changes are really good. The only concern I have is the word `Default`.  I don't think that's accurate in this case. Default implies something that will be used unless overridden (since that's how we use it in other cases in core)

`Active` or `Current` might be a better fit. I prefer `Set as Active` over `Set as Current` so that would be my pick. But we can debate this wording change after the PR is merged. 

@docwilmot are you still around and able to push what you have?

**Comment by jenlampton** [comment]
I'm going to take a stab at repeating @docwilmot's work to remove the node form changes...  this is so close I'd love to get it in!

**Comment by jenlampton** [comment]
I've reverted the change from the node form, and changed all instances of `Default revision` to `Active revision` in the PR. 

I did notice some strange behaviors when using the revisions UI though. Using the "Set as active" link does not set the current revision as active. Instead, it does what Backdrop did before, which is to copy that revision, and create a new one that is active.

When Editing a revision there are also other problems like a tab that says "Edit Draft" (don't we use the word `Draft` elsewhere for a different purpose?) and the Revision message textarea is not properly populated. To make matters more confusing, the option to create a new revision (from a current, unpublished revision) remains, and the unpublished revision shows the published checkbox as checked.

In short, editing revisions still needs a little cleanup. In order to get this in today, I think we should remove the ability to edit revisions, and restore the previous functionality.

**Comment by quicksketch** [comment]
> most important it adds the ability to edit any revision, and changes the wording from "Revert" which is a bit cryptic to me. Did we want to also remove all these changes from the PR?

Editing a revision can lead to a lot of potential issues. In the current PR, we can see issues such as:

- The "Published" radio button being selected when editing a previous revision. But it's not the "active" revision, so calling it "Published" is confusing, it's not accessible to the public.
- The tab at the top of the page is "Edit Draft", but this may have been a previous revision. Draft indicates something that hasn't been published yet.
- Editing a non-active revision doesn't repopulate the "Revision Log" field.
- And what happens when I check "Create new revision" off an old revision? It seems like only the active revision should have the ability to create a new revision.

**Comment by quicksketch** [comment]
I left additional feedback on @jenlampton's new PR at https://github.com/backdrop/backdrop/pull/3032

On closer inspection I don't think I'm comfortable with the current state of this issue. And we're changing a lot of things at the last minute that @docwilmot may not be able to approve. I think it may be best to wait on this issue post feature-freeze.

**Comment by jenlampton** [comment]
Bumped to 1.16

**Comment by docwilmot** [comment]
> changed all instances of Default revision to Active revision in the PR.

Bikeshed: `Active revision` sounds like the one you're working on now, like the one youre editing, or saving or viewing. I prefer "Default". But I'm not married to it.

**Comment by stpaultim** [comment]
I don't know if this is helpful. But, I just got an inside peek at a Django site and was interested in how they manage this. I like some of this UI, but don't know if any of it would work for us. But, it can't hurt to look and seek inspiration.

![revisions-django](https://user-images.githubusercontent.com/3144571/73113699-f0aec680-3eda-11ea-83a3-514de789c270.png)

The screen shot is an admin view of a blog post. If you click on the edit button you can scroll down and edit specific elements on this page. The page consists of several blocks (think paragraphs in Drupal world). Each block is edited separately and the page is either saved or published. 

#1) Toggles visibility of the list of revisions.
#2) The left and right arrows will navigate you through the existing revisions.
#3) You can publish any revision at anytime
#4) If you toggle/click the EDIT button on any revision - a new unsaved revision will be created at the top of the list (cloning the content from whichever revision you were viewing at the time) and a "SAVE" button will appear here. 

* You can edit any revision - create a new revision based upon the revision you wish to edit.
* You can only save the latest revision. 
* You can published any revision at any time.

**Comment by docwilmot** [comment]
@stpaultim thanks, will work toward that UI suggestion now.

**Comment by stpaultim** [comment]
Well, I like parts of this. But, before you put too much work into it, I think it would be helpful to hear what others think?

**Comment by docwilmot** [comment]
Slightly modified PR after all:
- If you do node edits through the node edit form, everything works just like before this PR. If you selected new revision, that revision is always set as the default revision.
- On the revision table though, now whenever you click EDIT on a revision which is older than the default revision, you will automatically create a new revision. This new revision will not be the default though.
- If you edit a revision which is more recent than default, you have the option to either just save, or create a new revision. This new revision will also not be the default.
- So to make a forward or older revision the new default, you need to set this through the revisiontable dropbutton.

I think this is a reasonable compromise that keeps the new functionality relatively unobtrusive.

ps I revived my old PR rather than continuing from @jenlampton's because I am not sure about Active vs Default as the right terminology. I think Default keeps Drupal compatibility better and is lesss confusing.
For some reason though I was unable to reopen the old PR so pushed a new one.

**Comment by stpaultim** [comment]
@docwilmot - Great work. Seeing this helps me clarify what I liked about the Django version of this. 
1) I know now, that I don't like "default." That just does't make sense for me. I like "Published Version" best. I like "active" better than "default," but I understand @docwilmot's concern. 

2) I like "Published version" or "Active version" rather than "Published revision" or "Active revision" 

3) I like something like this, in terms of how the published version is displayed.
![Versions](https://user-images.githubusercontent.com/3144571/73244318-413e5200-416f-11ea-83b3-987a40b0149f.png)

**Comment by docwilmot** [comment]
> I know now, that I don't like "default." ... I like "Published Version" best. I like "active" better than "default," but I understand your concern.

The difficulty we have is that "published" means something completely different in Backdrop; any revision, forward or otherwise can be published or unpublished, so we cant do that. Or can we?
> I like something like this, in terms of how the published version is displayed.

I like the idea of automatically making a comment about which revision is the parent revision; I think we should adopt that.
> I do not think that the numbers are helpful if they are cumulative. 

Those are actually the revision IDs. Agreed theyre not terribly useful, but I think thats previous behaviour? Could get rid of it though and just use the node title. 
> I would think that you could publish an old version without making a new version

What do you mean by "publish"?
>It's not clear to me how to edit the current/published version and create an unpublished draft of it, keeping the current version in place. 

This was the whole holdup it seems. The functionality before this PR was that every time you created a new revision from the node form, it MUST become the default revision, so without a switch in the node form (which the PR reviewers are against) this is impossible to prevent this. You'll have to save a new, unchanged revision (just check the box for new revision and save), then go to the revision table, find the second to last revision (it would be identical to the just-created default revision) and edit that. 
I could also add a dropbutton link to the default revision in the revision table that says "create a new revision" to skip that step. But still all forward revisions would have to be created from the revision table regardless.
Hope you get this, as it sounds convoluted even to me typing it.
>To make a version published or active, "revert" is not the right word.

"Revert" should be gone? Where are you seeing this?

**Comment by laryn** [comment]
Thanks for all the work here! To throw another term for consideration out there, how does "Primary version" sound compared to "Default revision" or "Active version"? Otherwise "Current version" seems reasonable to me as well.
(And I know it's not a part of this PR, but as I was thinking about the "edit the current/published version and create an unpublished draft of it, keeping the current version in place" workflow -- I think being able to schedule new revisions to replace the current version at a specific point in the future would be amazing.)

**Comment by stpaultim** [comment]
@docwilmot - Thanks again.

> "Revert" should be gone? Where are you seeing this?

If you click on "Set as Default" for a revision, you get a message `Are you sure you want to revert to the revision from ...`

> The difficulty we have is that "published" means something completely different in Backdrop;

I may be missing some of the implications of `default`. I think I am using `publish or published` in the same way that you are currently using `default.` I suggest that `publish` is for the active version that is available to the public. `Published` is the current public version and `publish` is the act of making `published.` Isn't that how we already use it. Whichever version is `published` is available to the public (not a draft). You can only have one `published` version at a time and you might not have any.

Again, I could be missing some other aspects of how that term is currently used.

**Comment by laryn** [comment]
@stpaultim I think the confusion would come in when working on a page that is not published. Is there then no default revision?

**Comment by jenlampton** [comment]
There could be an `active revision` even when the page is not published.

**Comment by laryn** [comment]
There would have to be a default/active/primary revision in order for the publish function on the main node edit form to work, right?

**Comment by stpaultim** [comment]
> I think the confusion would come in when working on a page that is not published. Is there then no default revision?

I'm also confused about the difference between active/default and published. I don't understand the use case for an "active" or "default" status that is NOT published. Any time you edit a revision, a new revision is made. So, the most recently worked on revision is always (by definition) the top/recent revision. I would assume that the most recently worked on revision is the "active" one (whether it is published or not). 

If you have 6 revisions, numbered 1-6 in order of most recent changes.

#1 Oldest version
#2 Second oldest version
#3 Third oldest version 
#4 Fourth oldest version (PUBLISHED)
#5 Fifth oldest version - some changes after publishing *4*
#6 The most recent unpublished changes

If you edit #2, it automatically creates revision #7. You can not save changes to an old revision. So, by default the active version is the most recent version (I think). I don't understand the use case to designate version #3 (or any old version) as default or active if it is not published?

Does this make sense? What am I missing?

**Comment by jenlampton** [comment]
> If you edit #2, it automatically creates revision #7. You can not save changes to an old revision.

This is how it works currently, but I don't think this is how it works after the PR, I don't think? 

> So, by default the active version is the most recent version (I think).

This may also be how it currently works...  but IIRC this is what is changing?

I was using the word `active` to mean the one that currently displayed when anyone views the page (either published or not). You would not want the newest revision to automatically become the active one until it has been approved, so there needs to be newer / revisions that are getting edited while an older one is still active.

**Comment by laryn** [comment]
@stpaultim In the case of a page that is entirely unpublished but has multiple revisions, the term "published" on a revision wouldn't work -- or are you suggesting that unpublishing a page would then just mark the revision unpublished (like the rest of the revisions)? And then clicking publish on the page would then publish the most recent revision rather than the one that may have been published moments ago? Just trying to get my head around this...

**Comment by stpaultim** [comment]
> You may not want the newest revision to automatically become the active one until it has been approved (this is the use-case we are solving) so there needs to be able to newer revisions that are getting edited while an older one remains active.

OK, this helps me understand the use case and it does justify `active` as a possible term. I though the use case was to allow a more current revision than the one that is published. The idea that I would have a draft post waiting approval to be published that is not the most recently updated version seems like a pretty specialized use case. But, I get it. 

> This is how it works currently, but I don't think this is how it works after the PR. Every revision has an 'Edit' link.

In my testing, if you click "Edit" on an old revision, it creates a new revision. You can not edit and save an old revision. Which makes sense to me. 

>or are you suggesting that unpublishing a page would then just mark the revision unpublished (like the rest of the revisions)?

Yes, I am suggesting that an unpublished page would have NO published versions and a published page would always have ONE published version, this feature would allow you to say which which revision is displayed on a published node. 

I could be misunderstanding the original intention. But, what I am describing makes sense to me and is very close to what is being used in Django.

**Comment by jenlampton** [comment]
I feel like the word `Published` should maybe only be applied to the node as a whole. A node is either published, or it's not. The revisions should maybe have some other label, like `Approved` or something? If a node is published, there must be at least one approved revision. And if there is more than one approved revision, than one of those should be marked `active` (or `currently displayed` or something that people will understand)

**Comment by stpaultim** [comment]
@jenlampton - I think your description is one extra layer of functionality. It's not a bad idea, I just find it more complicated and pretty specialized. 

The Django model felt really useful to me and easy to understand - and is essentially what I'm describing. Adding an "approved" but "unpublished" status feels like a fringe case for advanced editor workflows.

**Comment by laryn** [comment]
I think I'm envisioning something very similar to you, @stpaultim, but the only potential hitch I see is in the "published" terminology. In the case I described, you could get wildly different content simply by unpublishing a page and then publishing again if there was no way to mark revisions as the current/active revision.

This makes sense to me right now:

- Have a way to mark revisions as "current/active/whatever" (and on the revisions tab perhaps also indicate the node status -- e.g. "active revision (published)";  "active revision (unpublished)"
- So when unpublishing a node, the "active" flag is never lost and republishing will republish the same revision that was published last -- unless manually changed.

**Comment by docwilmot** [comment]
For clarity, the Node object in Backdrop and Drupal has a property called `Node->status` and a corresponding `status` column in the database which defines whether it is "published" or not. `Status = 0` is unpublished. The status is set by triggering the published/unpublished/scheduled radios. This is completely independent of the revision system. So any revision can be set as one of those options whether default or not. 

It would be great if we could deprecate that and just use forward/backward revisions, but status is too ingrained in core and contrib, so we couldnt remove this now. We need therefore to work around the terminology.

**Comment by jenlampton** [comment]
> In my testing, if you click "Edit" on an old revision, it creates a new revision. You can not edit and save an old revision. Which makes sense to me.

Well, then we'll need a verb other than `Edit`. If I click a link that says `Edit` I expect that I would be editing that thing. Maybe `Clone` would be better here?

> this feature would allow you to say which which revision is displayed on a published node.

So would having a single `active` revision. We might be saying the same thing here, but with different words... I am trying to avoid `Published` since that already has a meaning that's not quite the same as this.

> @jenlampton - I think your description is one extra layer of functionality. It's not a bad idea, I just find it more complicated and pretty specialized.

You may be right, I'm thinking about workbench / workflow (but without additional states) since that UI made the most sense to me. 

> This [note published status] is completely independent of the revision system.

Oh good, so we can safely use a different word than `Published` without adding too much confusion :) 

> It would be great if we could deprecate that and just use forward/backward revisions

I don't think that would be better, but fortunately we don't need to worry about it in 1.x. We can debat

> So any revision can be set as one of those options whether default or not.

Is the `status` column in the node table or in the node_revision table? (or in both?)

If the node itself had a status, but **not the revisions** then a revision wouldn't be published, unpublished, or scheduled, but could instead have an `Active` / `not active` status which might make a lot more sense.

**Comment by docwilmot** [comment]
> Well, then we'll need a verb other than Edit. If I click a link that says Edit I expect that I would be editing that thing. Maybe Clone would be better here?

I like `Edit` personally, but lets say Clone. We could add this clone link to the Active (I'm suspecting everyone prefers "Active", so I'm jumping on) and also to all *older* revisions. Then launch the node form when you click that. That would also then solve @stpaultim's wish for a way to get a forward revision from the Active node. 

>Is the status column in the node table or in the node_revision table? (or in both?)

Actually Node, but when you click to view a revision, it must be loaded as a Node, to get all the fields and whatnot.

We could simply set all revisions automatically to `status = 1` (meaning "published, in the current pre-PR meaning) but then hide the status fieldset from the node form. So the publishing UI would only be available to the Active revision's node form.

Either way we're still left with the confusion that the Active revision can still be "unpublished". I dont know how we would address that.

**Comment by docwilmot** [comment]
@klonos @stpaultim can you have another look please? https://github.com/backdrop/backdrop/pull/3051. Test fails are expected, no point fixing them till we're sure about the code and UI changes.

**Comment by jenlampton** [comment]
@klonos @stpaultim I have tried to document the results of our conversation (along with agreed-upon terminology) in this google doc: http://bit.ly/37TuoDi 

Once you two have a chance to review that, we can either put it in this issue as a comment or summary update, or maybe make a new issue?

**Comment by docwilmot** [comment]
@klonos @stpaultim its probably unfair to ask this but I think the document was meant to summarize the team consensus, but reading it now sounds like multiple approaches, not consensus. Here's the unfair bit: can we ask the team to decide on an approach? Happy to code the decision.

**Comment by stpaultim** [comment]
@docwilmot - You are correct. The document was supposed to reflect consensus, but it doesn't yet. Did you see the portion of the last dev meeting when we talked about this?
 
You could delete the section about underlying technical goals and you would pretty much have Jen's attempt to convey the consensus from our call. However, to some extent the consensus came from my own inability to keep up with the conversation. I still have some concerns, but they are not severe concerns and I do not want to hold up the process. 

I would like to look at the document again (after our meeting on Thurs) and see if I could integrate my ideas into what Jen wrote, but I'm not sure that will happen in the next day or two. 

The "Underlying Techincal Goals" was INTENDED to be my interpretation of what this issue was trying to achieve, whether or not it is actually in the PR. Part of what I was looking for was this feedback from you. ` This section though goes beyond the scope of this PR.`

From my perspective, I am willing to keep participating in the discussion and trying to better articulate what the short and long term goals are for this issue OR just step back and let you work on what Jen articulated and then we can iterate in the future. 

I can't really contribute much with the current status of the issue, because I'm still trying to process what our ultimate goals are.

**Comment by BWPanda** [comment]
If Backdrop features had code names, this one could be **'Backdraft'** :wink: :smile:

**Comment by stpaultim** [comment]
OK, here is another (slightly opinionated) attempt to summarize this issue:
https://docs.google.com/document/d/1bZ2i5FboKLYjKNeR1aiTtvyqarNmfLD9IkBIDwbsh1g/edit#

@jenlampton @klonos @docwilmot

**Comment by klonos** [comment]
@jenlampton @stpaultim @laryn and myself are currently working on finalizing the terminology to be used + the general specs of how this should work. To that end, I'm removing the "needs CR" and "needs testing" labels from this issue.

We're close to a decision! 🙂

**Comment by docwilmot** [comment]
Please remind me, was there a decision about what to do here?

**Comment by stpaultim** [comment]
Because this issue has verged dramatically since it was opened and the conversation gone in many directions, we had a conference call to discuss options and come up with a clearer status/plan. 

I've opened this new issue for the forward revisions portion: 
https://github.com/backdrop/backdrop-issues/issues/4354

I also opened this issue. However, this one needs a lot of work because I don't have nearly as good an understanding of the details of this issue. 

Better/advanced Workflow States
https://github.com/backdrop/backdrop-issues/issues/4355
